coderd/prebuilds/preset_snapshot.go

```diff
@@ -0,0 +1,256 @@
+// - Eligible: Number of prebuilds that are ready to be claimed
+// - Extraneous: Number of extra running prebuilds beyond the desired count
+// - Starting/Stopping/Deleting: Counts of prebuilds in various transition states
+//
+// The function takes into account whether the preset is active (using the active template version)
+// and calculates appropriate counts based on the current state of running prebuilds and
+// in-progress transitions. This state information is used to determine what reconciliation
+// actions are needed to reach the desired state.
+func (p PresetSnapshot) CalculateState() *ReconciliationState {
+	var (
+		actual     int32
+		desired    int32
+		eligible   int32
+		extraneous int32
+	)
+
+	if p.isActive() {
+		// #nosec G115 - Safe conversion as p.Running slice length is expected to be within int32 range
+		actual = int32(len(p.Running))
+		desired = p.Preset.DesiredInstances.Int32
+		eligible = p.countEligible()
+		extraneous = max(actual-desired, 0)
+	}
+
+	starting, stopping, deleting := p.countInProgress()
+
+	return &ReconciliationState{
+		Actual:     actual,
+		Desired:    desired,
+		Eligible:   eligible,
+		Extraneous: extraneous,
+
+		Starting: starting,
+		Stopping: stopping,
+		Deleting: deleting,
+	}
+}
+
+// CalculateActions determines what actions are needed to reconcile the current state with the desired state.
+// The function:
+// 1. First checks if a backoff period is needed (if previous builds failed)
+// 2. If the preset is inactive (template version is not active), it will delete all running prebuilds
+// 3. For active presets, it calculates the number of prebuilds to create or delete based on:
+//   - The desired number of instances
+//   - Currently running prebuilds
+//   - Prebuilds in transition states (starting/stopping/deleting)
+//   - Any extraneous prebuilds that need to be removed
+//
+// The function returns a ReconciliationActions struct that will have exactly one action type set:
+// - ActionTypeBackoff: Only BackoffUntil is set, indicating when to retry
+// - ActionTypeCreate: Only Create is set, indicating how many prebuilds to create
+// - ActionTypeDelete: Only DeleteIDs is set, containing IDs of prebuilds to delete
+func (p PresetSnapshot) CalculateActions(clock quartz.Clock, backoffInterval time.Duration) (*ReconciliationActions, error) {
+	// TODO: align workspace states with how we represent them on the FE and the CLI
+	//	     right now there's some slight differences which can lead to additional prebuilds being created
+
+	// TODO: add mechanism to prevent prebuilds being reconciled from being claimable by users; i.e. if a prebuild is
+	// 		 about to be deleted, it should not be deleted if it has been claimed - beware of TOCTOU races!
+
+	actions, needsBackoff := p.needsBackoffPeriod(clock, backoffInterval)
+	if needsBackoff {
+		return actions, nil
+	}
+
+	if !p.isActive() {
+		return p.handleInactiveTemplateVersion()
+	}
+
+	return p.handleActiveTemplateVersion()
```

Comment:
suggestion: I think you could inline these methods. Also, `CalculateState()` is called by both `handleInactiveTemplateVersion` and `handleActiveTemplateVersion`, which doesn't make sense to me. Shouldn't this function be working off a single snapshot of the state?

Reply:
@johnstcn when you say `inline`, you mean merge source code from `handleInactiveTemplateVersion`, `handleActiveTemplateVersion` and potentially `needsBackoffPeriod` into one function?
It was done like this before: https://github.com/coder/coder/commit/fe60b569ad754245e28bac71e0ef3c83536631bb#diff-d3a6533fdc6d2caf1cd5f5bb80f34eed89b3c68157249999b0fffb96af9fde05R13-R167.

But it was a bit hard for me to track the logic, so I implemented it this way.

Reply:
I kind of agree with your 2nd question, let me explain why it was done like this.

We have such composition of objects:
- global snapshot contains global database snapshot (data directly from SQL queries). Kind of raw data.
- preset snapshot is subset of filtered data from global snapshot. Also raw data.
- CalculateState is kind of derived/calculated data from preset snapshot.

We can try to merge `PresetSnapshot` & `ReconciliationState` into one object. Then we can call CalculateState inside `NewPresetSnapshot()` constructor. But not sure if it's better.